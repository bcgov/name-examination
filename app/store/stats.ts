import { getStats } from '~/util/namex-api'
import { getFormattedDateWithTime } from '~/util/date'
import { sortNameChoices } from '~/util'
import { type NameRequest } from '~/types'
import { defineStore } from 'pinia'

export const useStatsStore = defineStore('stats', () => {
  /**
   * Fetches paged statistics data based on the provided timespan and myStats flag.
   *
   * @param {string | null} timespan - The timespan for the statistics data. If null, the default timespan is used.
   * @param {boolean} myStats - A flag indicating whether to fetch statistics related to user's own stats.
   * @return {Promise<any>} A promise that resolves with the fetched statistics data.
   * @throws {Error} Throws an error if the fetching process fails.
   */
  async function getPagedStats (timespan: string | null, myStats: boolean): Promise<any> {
    const params = new URLSearchParams({
      timespan: timespan || '1',
      currentpage: '1',
      perpage: '1000'
    })
    if (myStats) params.append('myStats', 'True')

    try {
      const response = await getStats(params)

      if (response.ok) {
        const jsonData = await response.json() // Parse response JSON

        if (jsonData.numRecords >= 0) {
          const requestsData = jsonData.nameRequests as Array<NameRequest>

          const processedData = requestsData.map((statsData:any) => {
            statsData.names = sortNameChoices(statsData.names)
            if (statsData.comments.length === 1 && statsData.comments[0].timestamp) {
              statsData.comments[0].timestamp = getFormattedDateWithTime(statsData.comments[0].timestamp)
            }
            statsData.lastUpdate = getFormattedDateWithTime(statsData.lastUpdate)
            statsData.names.forEach((name:any) => {
              if (name.decision_text) {
                name.decision_text = name.decision_text.split(/\n\n/)
              }
            })
            return statsData
          })
          return {
            nameRequests: processedData,
            numRecords: jsonData.numRecords
          }
        }
      } else {
        throw new Error('Network response was not ok')
      }
    } catch (error) {
      console.error('There was a problem with your fetch operation:', error)
      throw error // Rethrow the error to handle it elsewhere if needed
    }
  }

  return {
    getPagedStats
  }
})
