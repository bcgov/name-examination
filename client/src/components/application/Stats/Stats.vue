<!--eslint-disable-->
<template>
    <div>
      <div class="container-fluid">
        <div class="row">
          <div class="col">
            <form class="form-inline" @submit.prevent="onGetStatsClick" >
              <input class="form-control" type="number" placeholder="hrs"
                     v-model="timespan" style="width: 100px;">
              <button class="btn btn-success btn-sm" style="margin-left: 10px;" type="submit">Get Stats</button>
            </form>
            <div class="font-italic" style="width: 100px; text-align: center;">hours</div>
          </div>

          <div class="col">
            <div class="float-right">Decisions in the past {{ timespanHuman }}: {{ numRecords }}</div>
          </div>
        </div>
<br>
        <div class="row">
          <div class="col">
            <b-pagination v-on:input="pageChanged" size="md" :total-rows="numRecords" :current-page="currentPage" :v-model="currentPage" :per-page="perPage">
            </b-pagination>
          </div>
        </div>


        <div class="b-table">
          <b-table striped hover
                   show-empty
                   responsive
                   ref="table"
                   :items="getPagedStats"
                   :fields="fields"
                   :current-page="currentPage"
                   :per-page="perPage"
                   :v-model="requests"
          >
            <span slot="names" slot-scope="data" v-html="data.value"></span>
            <template slot="nrdetails" slot-scope="data">
              <span class="add-top-padding"><b>{{data.item.nrNum}}</b></span>
              <span class="row">
                  <ol class="add-top-padding">
                  <li v-for="name in data.item.names">
                   {{ name.name }}
                    <span v-if="name.state != 'NE'" class="decision" :class="name.state">{{ name.state }}</span>
                    <p class="decision-reason">{{ name.decision_text }}</p>
                  </li>
                </ol>
              </span>
               <div class="row">
              <div class="col comment" v-if="data.item.comments.length">
                <h3>Last Comment</h3>
                <p>
                  <span class="comment-examiner">{{ data.item.comments[data.item.comments.length-1].examiner }}</span>
                   -
                  <span class="comment-timestamp">{{ new Date(data.item.comments[data.item.comments.length-1].timestamp).toLocaleString('en-ca',{hour:'2-digit',minute:'2-digit',day:'2-digit',month:'2-digit',year:'numeric'}) }}</span>
                </p>
                <p class="comment-text">{{ data.item.comments[data.item.comments.length-1].comment }}</p>
              </div>

            </div>
          </template>
          </b-table>
        </div>
     </div>
    </div>
</template>

<script defer>
/* eslint-disable */
import axios from '@/axios-auth'

export default {
  name: 'stats',
  data: function () {
    return {
      timespan: 1, // number of hours back to look at
      requests: null,
      numRecords: null,
      currentPage: 1,
      perPage: 50,
      totalRows: null,
      fields: [
        {
          key: 'nrdetails',
          label: "Request Details",
          _showDetails: true
        },
        {
          key: 'stateCd',
          sortable: true,
          label: 'Status'
        },
        {
          key: 'lastUpdate',
          sortable: true
        },
        {
          key: 'activeUser',
          sortable: false
        },
        {
          key: 'furnished',
          label: "Sent back to NRO?"
        }
      ]
    }
  },
  computed: {
    timespanHuman() {

      var retval = '';
      var tempTimespan = this.timespan;

      // days
      if (tempTimespan >= 24) {
        var days = Math.floor(tempTimespan / 24);
        retval += days + ' days ';
        tempTimespan = tempTimespan - (days * 24);
      }

      // hours
      if (tempTimespan >= 1) {
        retval += tempTimespan + ' hours ';
      }

      return retval;
    }
  },
  methods: {
    onGetStatsClick(e) {
        this.$refs.table.refresh();
    },
    getPagedStats(ctx) {
      
        this.requests = null;
        if (this.currentPage === 0) this.currentPage = 1

        var params = '';
        if (this.timespan != '' && this.timespan != null) {
          params = '?timespan=' + this.timespan;
        } else {
          params = '?timespan=1'
        }

        let currentPage = ctx.currentPage ? '&currentpage=' + ctx.currentPage : this.currentPage ? '&currentpage=' + this.currentPage : '';
        let pageSize = ctx.perPage ? '&perpage=' + ctx.perPage : this.perPage;
        let url = '/api/v1/requests/stats' + params + currentPage + pageSize;
        let promise = axios.get(url, {headers: {Authorization: `Bearer ${sessionStorage.getItem('KEYCLOAK_TOKEN')}`}});

        return promise.then(response => {

            if (response.data && response.data.numRecords > 0) {
                let requests = response.data.nameRequests;

                // sort names by choice number
                let r = requests.map((request)=> {
                    request.names = this.sortNames(request.names);
                    request.lastUpdate = this.convertDate(request.lastUpdate);
                    return request
                });

                this.numRecords = response.data.numRecords;
                return r
            }
        })
        .catch(error => console.log('ERROR: ' + error))

    },
    sortNames(data) {
      function compare(a,b) {
        if (a.choice < b.choice)
          return -1;
        if (a.choice > b.choice)
          return 1;
        return 0;
      }
      return data.sort(compare);
    },
    convertDate(thedate) {
        return new Date(thedate).toLocaleString('en-ca',{hour:'2-digit',minute:'2-digit',day:'2-digit',month:'2-digit',year:'numeric'});
    },
    pageChanged(page) {
      console.log("Page changed", page)
      this.currentPage = page
    }
  },
  watch: {
    currentPage: function(val) {
      console.log("CURRENT PAGE CHANGED: ", val)
    }
  }
}
</script>

<style scoped>

  .decision {
    font-weight: bold;
  }
  .decision.APPROVED, .decision.CONDITION {
    color: green;
  }
  .decision-reason {
    margin-left: 20px;
    white-space: pre-line;
  }
  .request-report {
    border-bottom: 1px solid darkgrey;
    margin-top: 10px;
  }
  .comment {
    padding-left: 40px;
  }
  .comment-text {
    white-space: pre-line;
  }
  .comment-timestamp, .comment-examiner {
    font-style: italic;
  }

</style>
