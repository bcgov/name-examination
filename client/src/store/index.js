/* eslint-disable */
import Vue from 'vue'
import Vuex from 'vuex'
import axios from '../axios-auth'
import globalAxios from 'axios'
import router from '@/router'

Vue.use(Vuex)

export default new Vuex.Store({
  state: {
    nrNum: null,
    authorized: null,
    userId: null,
    email: null,
    kctoken: null,
    nr_conflict: null,
    user_role: null,
    currentChoice: null,
    currentName: null,
    currentState: null,
    is_editing: false,
    nrData: null,
    details: null,
   // formData: {
      compInfo: {
        nrNumber: null,
        compNames: {
          compName1: null,
          compName2: null,
          compName3: null
        },
        requestType: null,
      },
      applicantInfo: {
        applicantname: {
          firstName: null,
          lastname: null
        },
        contactInfo: {
          address: null,
          contactName: null,
          phone: null,
          email: null,
          fax: null
        },
      },
      additionalCompInfo: {
        jurisdiction: null,
        natureOfBussiness: null,
        nuans: null,
        sk_name: null,
        nr_status: null
      },
      examiner: null,
      priority: null,
      reSubmission: {
        reSubmissionYN: null,
        linkedNR: null
      },
      reservationCount: null,
      expiryDate: null,
  //  },
    issueText: null,
    issue: {
        issue_Match: null,
        issue_Consent: null,
        issue_TradeMark: null,
        issue_History: null,
        issue_Format: null,
        issue_Match_Text: null,
        issue_Consent_Text: null,
        issue_TradeMark_Text: null,
        issue_History_Text: null,
        issue_Format_Text: null
      }
  },

  mutations: {
    authUser (state, userData) {
      state.kctoken = userData
      //state.kctoken = userData.client_session
      //state.userRole = userData.user_role
      //state.idToken = userData.client_session
    },

    clearAuthData (state) {
      state.kcToken = null
      state.userId = null
    },

    setIssueText (state,issue) {
      switch (issue) {
        case "M1":
          state.issueText = state.issue.issue_Match_Text;
          break;
        case "M2":
          state.issueText = state.issue.issue_Consent_Text;
          break;
        case "M3":
          state.issueText = state.issue.issue_TradeMark_Text;
          break;
        case "M4":
          state.issueText = state.issue.issue_History_Text;
          break;
        case "M5":
          state.issueText = state.issue.issue_Format_Text;
          break;
        default:
          state.issueText = state.issue.issue_Match_Text;
      }
    },

    loadpostgresNo(state, postgresData) {
        state.compInfo.nrNumber = postgresData.nameRequest
    },

    loadCompanyInfo(state, dbcompanyInfo) {
      console.log('Loading Company Info into state')

      state.nrData = dbcompanyInfo

      if (dbcompanyInfo.names.length == 0) {
        console.log('Error, No company names found')
        return
      } else {
        for (let record of dbcompanyInfo.names) {
          switch (record.choice) {
            case 1:
              state.compInfo.compNames.compName1 = record.name
              break;
            case 2:
              state.compInfo.compNames.compName2 = record.name
              break;
            case 3:
              state.compInfo.compNames.compName3 = record.name
              break;
            default:
              console.log('Error with company name structure')
              return
          }
        }
      }

      state.compInfo.requestType = dbcompanyInfo.requestTypeCd
      state.applicantInfo.applicantname.firstName = dbcompanyInfo.applicant
      //state.applicantInfo.applicantname.lastname = dbcompanyInfo.lastName
      //state.applicantInfo.contactInfo.address = dbcompanyInfo.address
      state.applicantInfo.contactInfo.contactName = dbcompanyInfo.contact
      state.applicantInfo.contactInfo.phone = dbcompanyInfo.phoneNumber
      //state.applicantInfo.contactInfo.email = dbcompanyInfo.email
      //state.applicantInfo.contactInfo.fax = dbcompanyInfo.fax
      state.additionalCompInfo.jurisdiction = dbcompanyInfo.xproJurisdiction
      state.additionalCompInfo.natureOfBussiness = dbcompanyInfo.natureBusinessInfo
      state.additionalCompInfo.nuans = dbcompanyInfo.nuansNum
      state.additionalCompInfo.sk_name = dbcompanyInfo.skPartner
      state.additionalCompInfo.nr_status = dbcompanyInfo.state
      state.examiner = dbcompanyInfo.userId
      state.priority = dbcompanyInfo.priorityCD
      //state.reSubmission.reSubmissionYN = dbcompanyInfo.resubmissionYN
      //state.reSubmission.linkedNR = dbcompanyInfo.linkedNR
      //state.reservationCount = dbcompanyInfo.reservationCount
      state.expiryDate = dbcompanyInfo.expiryDate
    },

    loadCompanyIssues(state, dbcompanyIssues) {
      //state.issue.issue_Match = dbcompanyIssues.isue_Match
      //state.issue.issue_Match_Text = dbcompanyIssues.issue_Match_Text
      //state.issue.issue_Consent = dbcompanyIssues.issue_Consent
      //state.issue.issue_Consent_Text = dbcompanyIssues.issue_Consent_Text
      //state.issue.issue_TradeMark = dbcompanyIssues.issue_TradeMark
      //state.issue.issue_TradeMark_Text = dbcompanyIssues.issue_TradeMark_Text
      //state.issue.issue_History = dbcompanyIssues.issue_History
      //state.issue.issue_History_Text = dbcompanyIssues.issue_History_Text
      //state.issue.issue_Format = dbcompanyIssues.issue_Format
      //state.issue.issue_Format_Text = dbcompanyIssues.issue_Format_Text
    },

    update_nrData(state){

      if(state.nrData.names.length == 0) {
        console.log('Error, No company names found')
        return
      } else {
        for (let record of state.nrData.names) {
          switch (record.choice) {
            case 1:
              state.nrData.names[0].name = state.compInfo.compNames.compName1
              break;
            case 2:
              state.nrData.names[0].name = state.compInfo.compNames.compName2
              break;
            case 3:
              state.nrData.names[0].name = state.compInfo.compNames.compName3
              break;
          }
        }
      }

      state.nrData.requestTypeCd = state.compInfo.requestType
      state.nrData.applicant = state.applicantInfo.applicantname.firstName
      //state.applicantInfo.applicantname.lastname = dbcompanyInfo.lastName
      //state.applicantInfo.contactInfo.address = dbcompanyInfo.address
      state.nrData.contact = state.applicantInfo.contactInfo.contactName
      state.nrData.phoneNumber =  state.applicantInfo.contactInfo.phone
      //state.applicantInfo.contactInfo.email = dbcompanyInfo.email
      //state.applicantInfo.contactInfo.fax = dbcompanyInfo.fax
      state.nrData.xproJurisdiction = state.additionalCompInfo.jurisdiction
      state.nrData.natureBusinessInfo = state.additionalCompInfo.natureOfBussiness
      state.nrData.nuansNum = state.additionalCompInfo.nuans
      state.nrData.skPartner = state.additionalCompInfo.sk_name
      //state.nrData.state =  state.additionalCompInfo.nr_status
      state.nrData.userId = state.examiner
      state.nrData.priorityCD = state.priority
      //state.reSubmission.reSubmissionYN = dbcompanyInfo.resubmissionYN
      //state.reSubmission.linkedNR = dbcompanyInfo.linkedNR
      //state.reservationCount = dbcompanyInfo.reservationCount
      state.nrData.expiryDate = state.expiryDate
    },

    saveDetail(state,detail){
      state.details = detail
    }
  },

  actions: {
    kcauth ({commit, dispatch, state}) {
      const kc = localStorage.getItem('KEYCLOAK_TOKEN')
      state.kctoken = kc
      //console.log(authData)
      //state.user_role =  authData.realm_access.roles,
      state.user_role = ''
      state.idToken = kc
      localStorage.setItem('kctoken',state.kctoken)
      localStorage.setItem('user_role',state.user_role)

      if (state.kctoken){
        console.log('KC Authorized user roles')
        commit('authUser', {
          //user_role: state.user_role,
          kctoken: state.kctoken
        })
      }
    },

    logout ({commit,state}) {

      commit('clearAuthData')

      localStorage.removeItem('expirationDate')
      localStorage.removeItem('token')
      localStorage.removeItem('user_role')
      localStorage.removeItem('userId')
      localStorage.removeItem('email')
      localStorage.removeItem('COMPINFO')

      localStorage.removeItem('kctoken')
      localStorage.removeItem('KEYCLOAK_TOKEN')
      localStorage.removeItem('AUTHORIZED')
      localStorage.removeItem('USERNAME')

    },

    tryAutoLogin ({commit}) {
    },

    storeUser ({commit, state}, userData) {
      if (!state.idToken) {
        return
      }
      globalAxios.post('/users.json' + '?auth=' + state.idToken, userData)
        .then(res => console.log(res))
        .catch(error => console.log(error))
    },

    fetchUser ({commit, state}) {
      if (!state.idToken) {
        return
      }
      globalAxios.get('/users.json' + '?auth=' + state.idToken)
        .then(res => {
          console.log(res)
          const data = res.data
          const users = []
          for (let key in data) {
            const user = data[key]
            user.id = key
            users.push(user)
          }
          console.log(users)
          commit('storeUser', users[0])
        })
        .catch(error => console.log(error))
    },

    setDetails({commit, state}) {
     var detail = state.details?null:"1";
     commit('saveDetail',detail)
    },

    selectNameIssue ({commit, dispatch, state}, divID) {
      console.log('action: selected issue control:' + divID)
      localStorage.setItem('issueText',divID )
      commit('setIssueText', divID)
    },

    getpostgrescompNo ({commit, dispatch, state}) {
      console.log('action: select next company number from postgres')
      const myToken = localStorage.getItem('KEYCLOAK_TOKEN')
      const url = '/api/v1/requests/queues/@me/oldest'
      const vm = this
      return axios.get(url, {headers: {Authorization: `Bearer ${myToken}`}}).then(response => {
        console.log('Comp No Response:' + response)
        commit('loadpostgresNo',response.data)
      })
    },

    getpostgrescompInfo ({commit},nrNumber) {
      console.log('action: getting data for company number: ' + nrNumber + ' from postgres')
      const myToken = localStorage.getItem('KEYCLOAK_TOKEN')
      const url = '/api/v1/requests/' + nrNumber
      console.log('URL:' + url)
      const vm = this
      return axios.get(url, {headers: {Authorization: `Bearer ${myToken}`}}).then(response => {
        console.log('Comp Info Response:' + response.data)
        localStorage.setItem('COMPINFO',response.data)
        commit('loadCompanyInfo',response.data)
      })
      .catch(error => console.log('ERROR: ' + error))
    },

    updateNRState ({state},nrState) {
      console.log('Updating Examination state for number: ' + state.compInfo.nrNumber)
      const myToken = localStorage.getItem('KEYCLOAK_TOKEN')
      const url = '/api/v1/requests/' + state.compInfo.nrNumber

      axios.patch(url,{headers: {Authorization: `Bearer ${myToken}`}},{"state": nrState} )
           .then(function(response){
                console.log('state updated to ' + nrState + ' for ' + state.compInfo.nrNumber)
            })
            .catch(error => console.log('ERROR: ' + error))
      },

    nameApproved( {commit, state}) {
      console.log('Name Approved for ' + state.compInfo.nrNumber + ", " + state.currentName)
      const myToken = localStorage.getItem('KEYCLOAK_TOKEN')
      state.nrData.state = 'APPROVED'
      commit('update_nrData')
      const url = '/api/v1/requests/' + state.compInfo.nrNumber
      axios.put(url, state.nrData, {headers: {Authorization: `Bearer ${myToken}`}})
           .then(function(response){
              console.log('Name approved for ' + state.compInfo.nrNumber)
           })
           .catch(error => console.log('ERROR: ' + error))
    },

    nameRejected({commit, state}) {
      console.log('Name Rejected for NR #' + state.compInfo.nrNumber)
      const myToken = localStorage.getItem('KEYCLOAK_TOKEN')
      state.nrData.state = 'REJECTED'
      commit('update_nrData')
      const url = '/api/v1/requests/' + state.compInfo.nrNumber
      axios.put(url,state.nrData, {headers: {Authorization: `Bearer ${myToken}`}})
           .then(function(response){
              console.log('Name rejected for ' +  state.compInfo.nrNumber)
            })
            .catch(error => console.log('ERROR: ' + error))
    }
  },

  getters: {
    is_editing(state) {
      return state.is_editing
    },
    user(state) {
      return state.user
    },
    email(state) {
      return state.email
    },
    issueText(state) {
      return state.issueText
    },
    isAuthenticated(state) {
      return localStorage.getItem("AUTHORIZED")
    },
    nrNumber(state) {
      return state.compInfo.nrNumber
    },
    compName1(state) {
     return state.compInfo.compNames.compName1
    },
    compName2(state) {
      return state.compInfo.compNames.compName2
    },
    compName3(state) {
      return state.compInfo.compNames.compName3
    },
    requestType(state) {
      return state.compInfo.requestType
    },
    firstName(state) {
      return state.applicantInfo.applicantname.firstName
    },
    lastName(state) {
      return state.applicantInfo.applicantname.lastname
    },
    address(state) {
      return state.applicantInfo.contactInfo.address
    },
    conEmail(state) {
      return  state.applicantInfo.contactInfo.email
    },
    contactName(state) {
      return state.applicantInfo.contactInfo.contactName
    },
    phone(state) {
      return state.applicantInfo.contactInfo.phone
    },
    fax(state) {
      return state.applicantInfo.contactInfo.fax
    },
    jurisdiction(state) {
      return  state.additionalCompInfo.jurisdiction
    },
    natureOfBusiness(state) {
      return state.additionalCompInfo.natureOfBussiness
    },
    nuans(state) {
      return state.additionalCompInfo.nuans
    },
    sk_name(state) {
      return state.additionalCompInfo.sk_name
    },
    nr_status(state) {
      return state.additionalCompInfo.nr_status
    },
    examiner(state) {
      return state.examiner
    },
    priority(state) {
      return state.priority
    },
    reSubmissionYN(state) {
      return state.reSubmission.reSubmissionYN
    },
    linkedNR(state) {
      return state.reSubmission.linkedNR
    },
    reservationCount(state) {
      return state.reservationCount
    },
    expiryDate(state) {
      return state.expiryDate
    },
    issue_Match(state) {
      return state.issue.issue_Match
    },
    issue_Match_Text(state) {
      return state.issue.issue_Match_Text
    },
    issue_Consent(state) {
      return state.issue.issue_Consent
    },
    issue_Consent_Text(state) {
      return state.issue.issue_Consent_Text
    },
    issue_TradeMark(state) {
      return state.issue.issue_TradeMark
    },
    issue_TradeMark_Text(state) {
      return state.issue.issue_TradeMark_Text
    },
    issue_History(state) {
      return state.issue.issue_History
    },
    issue_History_Text(state) {
      return state.issue.issue_History_Text
    },
    issue_Format(state) {
      return state.issue.issue_Format
    },
    issue_Format_Text(state) {
      return state.issue.issue_Format_Text
    },
    details(state) {
      return state.details
    },
    addInfo(state) {
      return state.details
    }
  }
})
