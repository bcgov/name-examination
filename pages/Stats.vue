<template>
  <div
    id="stats-viewer"
    class="mx-4 mt-4 flex h-[91vh] flex-col space-y-1"
  >
    <div class="stats-header">
      <h1 class="text-2xl font-bold text-gray-700">
        Statistics
      </h1>
    </div>
    <div class="flex items-center justify-between">
      &nbsp;
      <div class="font-semibold text-center">Total Examined Names: {{ numRecords }}</div>
      <div class="flex items-center space-x-2">
        <div class="flex items-center space-x-1">
          <input
            id="stats-checkbox"
            class="h-4 w-4"
            v-model="myStats"
            type="checkbox"
          >
          <label for="stats-checkbox" class="font-bold">My Stats</label>
        </div>
        <div class="timespan-input  align-baseline space-x-1">
          <label
            for="timespan"
            class="font-bold align-baseline"
          >Hours:&nbsp;</label>
          <TextInput
            id="timespan"
            ref="numberinput"
            v-model="timespan"
            type="number"
            class="!w-24"
          />
          <IconButton @click="fetchStats">
            Get Stats
          </IconButton>
        </div>
      </div>
    </div>

    <div class="h-full overflow-x-auto rounded-md border">
      <table class="min-h-fit min-w-full table-auto">
        <thead class="sticky top-0">
          <tr class="h-12 bg-bcgov-blue5 text-left text-sm text-white">
            <th
              v-for="(header, index) in fields"
              :key="index"
              class="border-b border-gray-200 px-2 py-1"
            >
              {{ header.label }}
            </th>
          </tr>
        </thead>
        <tbody
          v-if="!isLoading"
          class="text-sm"
        >
          <tr
            v-for="(item, index) in statsData"
            :key="index"
            class="align-top transition duration-200 ease-in-out border-b border-gray-300 hover:bg-gray-200"
          >
            <td class="whitespace-pre-line border-b border-gray-300 px-2 py-2">
              <div class="container-fluid">
                <div class="layout">
                  <div
                    class="flex my-2 font-bold"
                  >
                    {{ item.nrNum }}
                  </div>
                </div>
                <div
                  v-for="(name, i) in item.names"
                  :key="`names-layout-${i}`"
                >
                  <div
                    class="flex font-bold"
                  >
                    {{ i + 1 }}. {{ name.name }}
                      <span class="mx-2" :class="getClass(name.state)">
                        {{ name.state }}
                      </span>
                  </div>
                  <div class="layout">
                    <div
                      v-for="(text, j) in name.decision_text"
                      :key="`decision-text-flex-${j}`"
                      class="flex italic mb-2 ml-4"
                    >
                      {{ text }}
                    </div>
                  </div>
                </div>
                <div
                  v-if="item.comments.length > 0"
                  class="ml-2 mt-1 font-semibold"
                >
                  LAST COMMENT
                </div>
                <div
                  v-if="item.comments.length > 0"
                  class="dk-grey ml-2 mb-2"
                >
                  <div class="flex indent-10px">
                    {{ item.comments[0].comment }}
                  </div>
                  <div class="flex ft-ital indent-10px">
                    <template v-if="item.comments[0].examiner">
                      <template v-for="(value, name) in item.comments[0].examiner">
                        -{{ name }}: {{ value }},
                      </template>
                    </template>
                    <template v-else>
                      <span class="mr-2">â€“unknown examiner,</span>
                    </template>
                    {{ item.comments[0].timestamp }}
                  </div>
                </div>
              </div>
            </td>
            <td
              :class="getClass(item.stateCd)"
              class="font-semibold"
            >
              {{ item.stateCd }}
            </td>
            <td>{{ item.lastUpdate }}</td>
            <td>{{ item.activeUser }}</td>
            <td>{{ item.furnished }}</td>
          </tr>
        </tbody>
        <div
          class="absolute bottom-1/2 right-1/2 translate-x-1/2 translate-y-1/2 transform"
        >
          <LoadingSpinner v-if="isLoading" />
        </div>
      </table>
    </div>
  </div>
</template>

<script>
import { useStatsStore } from '~/store/stats.ts'
import { ref, onMounted } from 'vue'

export default {
  setup () {
    const timespan = ref('24')
    const numberinput = ref(null)
    const numRecords = ref(0)
    const myStats = ref(true)
    const statsData = ref([])
    const isLoading = ref(false)

    const stats = useStatsStore()

    const fields = [
      {
        key: 'nrdetails',
        label: 'Request Details',
        style: { width: '64%' }
      },
      {
        key: 'stateCd',
        label: 'Status',
        style: { width: '7%' }
      },
      {
        key: 'lastUpdate',
        label: 'Last Update',
        style: { width: '14%' }
      },
      {
        key: 'activeUser',
        label: 'Active User',
        style: { width: '10%' }
      },
      {
        key: 'furnished',
        label: 'Back to NRO?',
        style: { width: '5%' }
      }
    ]

    const fetchStats = async () => {
      try {
        isLoading.value = true

        const jsonData = await stats.getPagedStats(timespan.value, myStats.value)
        numRecords.value = jsonData.numRecords
        statsData.value = jsonData.nameRequests

        isLoading.value = false
      } catch (error) {
        console.error('Error fetching stats:', error)
      }
    }

    onMounted(async () => {
      await fetchStats()
    })

    /**
     * Determines the CSS class based on the provided state.
     * @param {string} state - The state value used to determine the CSS class.
     * @return {string | undefined} Returns the CSS class corresponding to the provided state. 
     * If the state does not match any condition, returns undefined.
     */
    function getClass (state) {
      if (state === 'APPROVED') {
        return 'approved'
      }
      if (state === 'REJECTED') {
        return 'rejected'
      }
      if (state === 'CONDITION' || state === 'CONDITIONAL') {
        return 'condition'
      }
    }

    return {
      fields,
      myStats,
      numRecords,
      timespan,
      statsData,
      numberinput,
      isLoading,
      getClass,
      fetchStats }
  }
}
</script>

<style scoped>
/* Add your scoped styles here */
.stats-header {
  padding: 3px;
  display: flex;
  align-items: center;
}

.checkbox-wrapper {
  position: relative;
  margin-right: 10px;
}

.timespan-input {
  display: flex;
  align-items: center;
}

.stats-table {
  margin-top: 10px;
}

.custom-button {
  background-color: #007bff; /* Blue background color */
  color: #000; /* Black text color */
  border: 1px solid #000; /* Black border */
  padding: 8px 16px; /* Adjust padding as needed */
  border-radius: 4px; /* Rounded corners */
  cursor: pointer; /* Show pointer cursor on hover */
}

.approved {
  color: #82c545
}

.rejected {
  color: #e01e36
}

.condition {
  color: #fcba19
}
/* Add more styles as needed */
</style>
