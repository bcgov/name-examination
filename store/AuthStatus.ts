import { defineStore } from 'pinia';

export const useAuthStore = defineStore('auth', {
  state: () => ({
    auth: false,
  }),
  actions: {
    toggleAuth() {
      this.auth = !this.auth;
    },
  },
});
