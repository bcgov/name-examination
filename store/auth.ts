import ConfigHelper from '../util/config-helper'
import KeycloakServices from '../public/keycloak/keycloak'
import { SessionStorageKeys } from '../util/constants'
import { defineStore } from 'pinia'

interface AuthState {
  token: string;
  idToken: string;
  refreshToken: string;
  kcGuid: string;
  loginSource: string;
}

export const useAuthStore = defineStore({
  id: 'auth',
  state: (): AuthState => ({
    token: '',
    idToken: '',
    refreshToken: '',
    kcGuid: '',
    loginSource: ''
  }),
  getters: {
    isAuthenticated (state): boolean {
      return !!state.token
    },
    keycloakGuid (state): string {
      return state.kcGuid || KeycloakServices.getUserInfo().keycloakGuid
    },
    currentLoginSource (state): string {
      return state.loginSource || KeycloakServices.getUserInfo().loginSource
    }
  },
  actions: {
    setKCToken (token: string): void {
      this.token = token
      ConfigHelper.addToSession(SessionStorageKeys.KeyCloakToken, token)
    },
    setIDToken (idToken: string): void {
      this.idToken = idToken
      ConfigHelper.addToSession(SessionStorageKeys.KeyCloakIdToken, idToken)
    },
    setRefreshToken (refreshToken: string): void {
      this.refreshToken = refreshToken
      ConfigHelper.addToSession(SessionStorageKeys.KeyCloakRefreshToken, refreshToken)
    },
    setKCGuid (kcGuid: string): void {
      this.kcGuid = kcGuid
    },
    setLoginSource (loginSource: string): void {
      this.loginSource = loginSource
    },
    clearSession (): void {
      // Reset all state properties related to the user's session
      this.token = ''
      this.idToken = ''
      this.refreshToken = ''
      this.kcGuid = ''
      this.loginSource = ''

      // Clear the session storage values
      ConfigHelper.removeFromSession(SessionStorageKeys.KeyCloakToken)
      ConfigHelper.removeFromSession(SessionStorageKeys.KeyCloakIdToken)
      ConfigHelper.removeFromSession(SessionStorageKeys.KeyCloakRefreshToken)
    },
    syncWithSessionStorage (): void {
      const token = ConfigHelper.getFromSession(SessionStorageKeys.KeyCloakToken) || ''
      const idToken = ConfigHelper.getFromSession(SessionStorageKeys.KeyCloakIdToken) || ''
      const refreshToken = ConfigHelper.getFromSession(SessionStorageKeys.KeyCloakRefreshToken) || ''

      this.token = token
      this.idToken = idToken
      this.refreshToken = refreshToken
    }
  }
})

if (import.meta.hot) {
  import.meta.hot.accept(acceptHMRUpdate(useAuthStore, import.meta.hot))
}
