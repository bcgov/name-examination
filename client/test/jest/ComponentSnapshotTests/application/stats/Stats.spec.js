/* eslint-disable */
import Vue from 'vue';
Vue.use(require('vue-shortkey'))
import { shallowMount } from '@vue/test-utils';
import Stats from '@/components/application/stats/Stats';
import store from '@/store'
import BootstrapVue from 'bootstrap-vue'
Vue.use(BootstrapVue)

describe('Stats.vue', () => {

    let component;
    beforeEach(() => {
        component = shallowMount(Stats, {store: store});
    });

   it("renders a Stats component", () => {
     expect(component.element).toMatchSnapshot();
   })

  // Add other tests specific to this component and not its sub-components
});
