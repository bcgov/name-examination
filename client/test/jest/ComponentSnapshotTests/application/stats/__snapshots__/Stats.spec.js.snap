// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stats.vue renders a Stats component 1`] = `
<div>
  <div
    class="container-fluid"
  >
    <div
      class="row"
    >
      <div
        class="col"
      >
        <form
          class="form-inline"
        >
          <input
            class="form-control"
            placeholder="hrs"
            style="width: 100px;"
            type="number"
          />
           
          <button
            class="btn btn-success btn-sm"
            style="margin-left: 10px;"
            type="submit"
          >
            Get Stats
          </button>
        </form>
         
        <div
          class="font-italic"
          style="width: 100px; text-align: center;"
        >
          hours
        </div>
      </div>
       
      <div
        class="col"
      >
        <div
          class="float-right"
        >
          Decisions in the past 1 hours : 
        </div>
      </div>
    </div>
     
    <br />
     
    <div
      class="row"
    >
      <div
        class="col"
      >
        <bpagination-stub
          current-page="1"
          perpage="50"
          size="md"
          v-model="1"
        />
      </div>
    </div>
     
    <div
      class="b-table"
    >
      <btable-stub
        apiurl=""
        currentpage="1"
        emptyfilteredtext="There are no records matching your request"
        emptytext="There are no records to show"
        fields="[object Object],[object Object],[object Object],[object Object],[object Object]"
        footvariant=""
        headvariant=""
        hover="true"
        items="function () { [native code] }"
        labelsortasc="Click to sort Ascending"
        labelsortdesc="Click to sort Descending"
        perpage="50"
        responsive="true"
        showempty="true"
        sortdirection="asc"
        striped="true"
        value=""
      />
    </div>
  </div>
</div>
`;
